### PR TITLE
SGD Optimiser, Dataloader, Power Operation and Successful Training !

### PR DESCRIPTION
**Pull request by karanchahal**
This is an exciting pull request ! We managed to train our first network on this framework !

This pull request adds:

1. SGD Optimiser along with the Base class for all optimisers.
2. The dataset for celsius to fahrenheit converter including the base class for all data loaders.
3. Introducing the power operation for our tensors to backpropagate on.
4. Tests for the above features.

And finally,

__A fully working demo of a small network learning how to become a celsius to fahrenheit converter !__

Onward and upward from now on ! Next stop: MNIST.

### CODE REVIEW

File: data/celsius2faranheit.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+*/
+
+#include "data/dataloader.h"
+#include <stdlib.h>
+
+#ifndef __C2F_DATASET_INCLUDED__   
+#define __C2F_DATASET_INCLUDED__ 
+
+template<typename I, typename T>
```
Comment (uditarora) [inline data/celsius2faranheit.h]: Do we need to support different variable types for Celsius and Fahrenheit temperatures?

File: data/celsius2faranheit.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/*
+    This file defines the Celsius To Faranheit DataLoader. It's input is variables containing the
```
Comment (uditarora) [inline data/celsius2faranheit.h]: nit: 'Fahrenheit' is misspelt. We should probably change it everywhere 🙈

File: buildTensorflow.h [hunk 3]
```diff
@@ -2,3 +2,5 @@
 #include "overloads/tensor.h"
 #include "operations/operations_Impl.h"
 #include "layers/dense.h"
+#include "optims/sgd.h"
+#include "data/celsius2faranheit.h"
```
Comment (uditarora) [inline buildTensorflow.h]: nit: newline at EOF

File: operations/operation.h [hunk 4]
```diff
@@ -40,6 +40,11 @@ class Operation {
     
     // New API for forward Prop
     virtual Tensor<T>* forward() = 0;
+
+    ~Operation() {
+        delete t1;
+        delete t2;
+    }
```
Comment (uditarora) [inline operations/operation.h:47]: What about t3?

File: optims/optim.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/*
+    This file defines the Base Class for all Optimizers.
+*/
+
+#include "types/tensor.h"
+#include "unordered_set"
```
Comment (uditarora) [inline optims/optim.h]: #include <unordered_set>?

File: optims/optim.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        for(auto i : params) {
+            i->zeroGrad();
+        }
+    }
+
+    // This overloaded function specifes how one optimisation step will be performed
+    virtual void step(T learning_rate) {};
+};
+
+#endif
```
Comment (uditarora) [inline optims/optim.h]: nit: newline at EOF

File: optims/sgd.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    It performs this update step by this formula:
+    
+    val = val - learning_rate*gradient_of_val
+    
+    Where val is the value of the tensor and gradient_of_val is the partial gradient of the
+    tensor with respect to the loss.
+*/
+
+#include "optims/optim.h"
+#include "queue"
```
Comment (uditarora) [inline optims/sgd.h]: #include <queue>

File: optims/sgd.h [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        step(this->lr);
+
+        // reset Gradients to zero
+        this->zeroGrad();
+       
+    }
+
+    // Performs 1 step of gradient descent. See top of the file to see definition of SGD. 
+    void step(T learning_rate) {
+
```
Comment (uditarora) [inline optims/sgd.h]: nit: remove extra newline

File: optims/sgd.h [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        for(auto t: this->params) {
+            t->val = t->val - learning_rate*t->grad;
+        }
+    }
+
+};
+
+#endif
```
Comment (uditarora) [inline optims/sgd.h]: nit: newline at EOF

File: overloads/vector.h [hunk 10]
```diff
@@ -62,6 +62,29 @@ vector<T> operator + (T a, const vector<T> &b) {
...
+    for(int i = 0;i<a.size();i++) {
+        T prod = a[i]-b[i];
+        arr.push_back(prod);
+    }
+    return arr;
+}
+
+// Scalar Subtraction
+template<typename T>
+vector<T> operator - (T a, const vector<T> &b) {
```
Comment (uditarora) [inline overloads/vector.h:79]: Should the arguments order be reversed? We would usually subtract a scalar from a vector, not vice-versa

File: data/celsius2faranheit.h (hunk 1 above)
Comment (karanchahal) [reply data/celsius2faranheit.h]: Hey ! Thanks for the review.
So not particularly required in this case, but the base Data loader class requires specifying the data types of inputs and targets and as it inherits from the Base class. I though I'd add it. I think I can remove it and hard code to float.

Should I do this ?

File: data/celsius2faranheit.h (hunk 2 above)
Comment (karanchahal) [reply data/celsius2faranheit.h]: yes, I noticed but was too lazy last night :P. Also I haven't written tests for the Dataloader. Have to do that too :(

File: operations/operation.h (hunk 4 above)
Comment (karanchahal) [reply operations/operation.h:47]: t3 will be deleted as it's a recursive call. So originally the user will write delete t3;
then the Destructor will go on and call delete t3->backOp.
which will delete t3->backOp->t1 and t3->backOp->t2.

Also, in C++ if you try to delete an already deleted object, it throws an error.

File: optims/optim.h (hunk 5 above)
Comment (karanchahal) [reply optims/optim.h]: Oh right

File: overloads/vector.h (hunk 10 above)
Comment (karanchahal) [reply overloads/vector.h:79]: Sometimes we need to do 1 - x. Or 1+ x, the sigmoid operation requires this.

Comment (karanchahal) [reply overloads/vector.h:79]: For example the backProp of sigmoid needs f*(1-f).

**Review by uditarora — APPROVED** [review]
